qiskit_experiments/framework/experiment_data.py

```diff
@@ -767,6 +770,81 @@ def add_data(
     ) -> None:
         """Add experiment data.
 
+        Args:
+            data: Experiment data to add. Several types are accepted for convenience:
+
+                * Result: Add data from this ``Result`` object.
+                * List[Result]: Add data from the ``Result`` objects.
+                * Dict: Add this data.
+                * List[Dict]: Add this list of data.
+
+        Raises:
+            TypeError: If the input data type is invalid.
+        """
+
+        if any(not future.done() for future in self._analysis_futures.values()):
+            LOG.warning(
+                "Not all analysis has finished running. Adding new data may "
+                "create unexpected analysis results."
+            )
+        if not isinstance(data, list):
+            data = [data]
+
+        # Directly add non-job data
+        with self._result_data.lock:
+            tmp_exp_data = ExperimentData()
```

Comment:
Can you make this more efficient? If data is not composite type, you don't need to instantiate new object, right? Instead of using `experiment_separator` dictionary, you can do

- check if "compoite_index" is in the metadata; if not, just append the datum to `self._result_data`. 
  - marginalize the datum
  - zip loop the composite_index and marginalized datum (they should have the same length)
    - get child experiment data of a composite index; if not exist (try-except), set new child data
    - store marginalized datum to the child data container, i.e. through the `add_data` method of the child container; this covers nested composite case because composite check is done recursively.